### DISCUSSION TITLE
validate onBlur not triggering - only on submit

### DISCUSSION THREAD

**Discussion by simplenotezy** [react-hook-form/react-hook-form#5138] [Q&A] [answered]
Just playing around with this library trying to get validation to work on onBlur instead of when submitting. Validation only happens when I click submit, not when I blur input, and really can't figure out why.

My Input component:

```
import classnames from 'classnames';
import { forwardRef } from 'react';
import { Root, InputWrapper, Label } from './Input.module.css';

const Input = forwardRef(
  (
    { children, className, label, name, htmlId, onChange, onBlur, ...props },
    ref
  ) => {
    const classes = classnames(Root, className);

    const allProps = props;
    const labelFor = allProps.id || name || htmlId;

    if (label && !allProps.id) {
      allProps.id = name || htmlId;
    }

    const input = (
      <input
        onChange={onChange}
        onBlur={onBlur}
        className={classes}
        {...allProps}
        ref={ref}
      />
    );

    return label ? (
      <div className={InputWrapper}>
        <label className={Label} htmlFor={labelFor}>
          {label}
        </label>
        {input}
        {children}
      </div>
    ) : (
      input
    );
  }
);
```

My form:

	const {
		register,
		formState: { errors },
		handleSubmit,
	} = useForm({
		mode: 'onBlur',
	});
	
	const onSubmit = (data) => console.log(data);

	<Input
		{...register('phone', { required: true, minLength: 10 })}
		type="text"
		label="Phone no."
		placeholder="+45"
	/>
       {errors.phone?.minLength && 'Min length'}

Any ideas why? I mostly just followed the Getting Started guide.

**Comment by bluebill1049** [comment]
can you provide a codesandbox?

**Comment by thanh-nguyen-95** [comment, marked as answer]
Since you're using a custom `<Input />` component, I recommend using `Controller` or `useController` for managing the field.

I did some testing and here's what I found:

- Validation still works as intended (`onBlur` event)
- Error messages are all *empty string*

Working CodeSandbox using latest version (7.5.2): https://codesandbox.io/s/cranky-wozniak-14z1u?file=/src/App.js 

You can either downgrade `react-hook-form` or specify the rules with both `value` and `message`

```js
{
  required: {
    value: true,
    message: 'Input is required'
  }
}
```

**Reply by simplenotezy** [reply]
Hi @thanh-nguyen-95 - thanks for your answer. I'll have a look at Controller. However, I am already using `7.5.2`

**Reply by thanh-nguyen-95** [reply]
> Hi @thanh-nguyen-95 - thanks for your answer. I'll have a look at Controller. However, I am already using `7.5.2`

I think the bug appears in recent versions, that's why I said you should downgrade to a lower version.

However, if you specify the rules as I did in the CodeSandbox, you can still make it work on latest version.

**Reply by bluebill1049** [reply]
Thanks @thanh-nguyen-95 

> I think the bug appears in recent versions, that's why I said you should downgrade to a lower version.

what's the bug?

**Reply by thanh-nguyen-95** [reply]
> Thanks @thanh-nguyen-95
> 
> > I think the bug appears in recent versions, that's why I said you should downgrade to a lower version.
> 
> what's the bug?

Validation triggered on `onBlur` event returns empty string as error message if you don't specify the `message` in `rules`. You can check my CSB

**Reply by simplenotezy** [reply]
@thanh-nguyen-95 I believe I had issues with errors in general, I was looking at the overall `error` object and got nothing. However, when using Controller it works as intended

**Reply by bluebill1049** [reply]
thanks @thanh-nguyen-95 how can i reproduce the issue in this CSB https://codesandbox.io/s/gallant-frog-yu6li?

**Reply by thanh-nguyen-95** [reply]
> thanks @thanh-nguyen-95 how can i reproduce the issue in this CSB https://codesandbox.io/s/gallant-frog-yu6li?

Submit the form and check the errors returned. The message is empty even though other properties work as intended

**Reply by bluebill1049** [reply]
thanks, I am investigating.

**Reply by bluebill1049** [reply]
hmmm i couldn't reproduce the bug: https://codesandbox.io/s/gallant-frog-yu6li?file=/src/App.js i have put a console inside the controller, let me know if this is working as epexected or perhaps you want to share a video clip? much appreciated @thanh-nguyen-95

**Reply by thanh-nguyen-95** [reply]
> 
> 
> hmmm i couldn't reproduce the bug: https://codesandbox.io/s/gallant-frog-yu6li?file=/src/App.js i have put a console inside the controller, let me know if this is working as epexected or perhaps you want to share a video clip? much appreciated @thanh-nguyen-95

My apology. Since I've always used the `{ value, message }` syntax for `rules`, I find it weird when `error.message` is an empty string. I thought there would be some sort of default error message there.

Everything is working as expected. Using `Controller` solved the problem

**Reply by bluebill1049** [reply]
no worries @thanh-nguyen-95 thank you for your support and help on everyone. ❤️ 🙏

**Comment by avremel** [comment]
One of the props which `register` returns is `name`. Both `name` and `ref` need to be passed through as props to `<input />` for `ref` to work ([docs](https://react-hook-form.com/api/useform/register)). Your `Input` component is destructuring the `name` prop but not passing it along to `<input />`.

The library code apparently relies on the `name` prop being present.

In this code example, if you remove the name prop, onBlur gets triggered correctly.
[Stackblitz demo](https://stackblitz.com/edit/react-nditke?file=src/App.js)
